rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -24,12 +24,123 @@
+            != sparse_sentence_features.features.shape[0]
+        ):
+            raise ValueError(
+                f"Sequence dimensions for sparse and dense sentence features "
+                f"don't coincide in '{message.get(TEXT)}'"
+                f"for attribute '{attribute}'."
+            )
+
+    out = {}
+    if sparse_sentence_features is not None:
+        out[(True, SENTENCE)] = sparse_sentence_features.features
+    if sparse_sequence_features is not None:
+        out[(True, SEQUENCE)] = sparse_sequence_features.features
+    if dense_sentence_features is not None:
+        out[(False, SENTENCE)] = dense_sentence_features.features
+    if dense_sequence_features is not None:
+        out[(False, SEQUENCE)] = dense_sequence_features.features
+
+    return out
+
+
+def extract_attribute_features_from_all_messages(
+    messages: List[Message],
+    attribute: Text,
+    type: Optional[Text] = None,
+    featurizers: Optional[List[Text]] = None,
+) -> Tuple[List[FeatureArray], List[FeatureArray]]:
+    """Collects and combines the attribute related features from all messages.
+
+    Args:
+        messages: list of messages to extract features from
+        attribute: the only attribute which will be considered
+        type: if not None, the other type (sentence/sequence) will be ignored
+        featurizers: the featurizers to be considered
+    Returns:
+        Sequence level features and sentence level features. Each feature contains
+        FeatureArrays with sparse features first.
+    """
+    if (type is not None) and (type not in [SENTENCE, SEQUENCE]):
+        raise ValueError(f"Unknown type {type}")
+    # for each label_example, collect sparse and dense feature (matrices) in lists
+    collected_features: Dict[
+        Tuple[Text, Text], List[Union[np.ndarray, scipy.sparse.spmatrix]]
+    ] = dict()
+    for msg in messages:
+        sparse_type_to_feature = extract_attribute_features_from_message(
+            message=msg, attribute=attribute, featurizers=featurizers,
+        )
+        for (feat_is_sparse, feat_type), feat_mat in sparse_type_to_feature.items():
+            if feat_type not in [SEQUENCE, SENTENCE]:
+                raise ValueError(f"Unknown type {feat_type}")
+            if (type is None) or (type == feat_type):
+                collected_features.setdefault((feat_is_sparse, feat_type), []).append(
+                    feat_mat
+                )
+
+    # finally wrap the lists of feature_matrices into FeatureArrays
+    # and collect the resulting arrays in one list per type:
+    sequence_features = []
+    sentence_features = []
+    for type, collection in [
+        (SEQUENCE, sequence_features),
+        (SENTENCE, sentence_features),
+    ]:
+        for is_sparse in [True, False]:
+            # Note: the for loops make the order explicit, which would
+            # otherwise (i.e. iteration over collected_features) depend on
+            # insertion order inside _extract_features

```

Comment:
I would add a line here saying how is it useful downstream in the code.
I guess `batch_to_model_data_format` will depend on this order?

Reply:
I think the order  inside the sub-sequences containing only sparse feature and dense features is critical. Haven't found a place yet where this mattered. The way the label_data was extracted with the "[0]" looked dangerous which was why I had added this.

Reply:
> Haven't found a place yet where this mattered. The way the label_data was extracted with the "[0]" looked dangerous which was why I had added this.

Okay I need to come back to this. I don't think I have the brain space to search for this right now but I promise I'll come back to it 😅

Reply:
I think that for the usage with "[0]" order won't matter because in that case we're using the default label features (only sparse sentence features (should be) in the features list then) 

The reason I'd still keep the function as is now: I find the function much easier to parse now than before. And we get the "sparse first" as a "bonus" so we can mention that in the docstring as well.